linux/hid.c

```diff
@@ -130,6 +135,48 @@ static wchar_t *utf8_to_wchar_t(const char *utf8)
 	return ret;
 }
 
+
+/* Set the last global error to be reported by hid_error(NULL).
+ * The given error message will be copied (and decoded according to the
+ * currently locale, so do not pass in string constants).
+ * The last stored global error message is freed.
+ * Use register_global_error(NULL) to indicate "no error". */
+static void register_global_error(const char *msg)
+{
+	if (last_global_error_str)
+		free(last_global_error_str);
+
+	last_global_error_str = msg? utf8_to_wchar_t(msg): NULL;
+}
+
+
+/* Set the last error for a device to be reported by hid_error(device).
+ * The given error message will be copied (and decoded according to the
+ * currently locale, so do not pass in string constants).
+ * The last stored global error message is freed.
+ * Use register_device_error(device, NULL) to indicate "no error". */
+static void register_device_error(hid_device *dev, const char *msg)
+{
+	if (dev->last_error_str)
+		free(dev->last_error_str);
+
+	dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
+}
+
+/* See register_device_error, but you can pass a format string into this function. */
+static void register_device_error_format(hid_device *dev, const char *format, ...)
+{
+	va_list args;
+	va_start(args, format);
+
+	char msg[100];
+	vsnprintf(msg, sizeof(msg), format, args);
+
+	va_end(args);
+
+	register_device_error(dev, msg);
+}
```

Comment:
Can someone confirm that this C code is written correctly and I didn't make some fundamental mistake with how the `char msg[100]` is handled? :D

Reply:
LGTM

Reply:
You pass the pointer to local `msg[100]` which is on stack, but in `register_device_error()` it is copied to `dev->last_error_str` which is dynamically allocated and it's lifespan is the same as lifespan of a device. This is copied before the stack is freed. So it's fine ;)

Reply:
Great! :)